File: Python/programmers/etc/level3/고고학_최고의_발견.py
import sys
INF = sys.maxsize

from itertools import product

dx, dy = [-1, 1, 0, 0, 0], [0, 0, 0, -1, 1]

def solution(clockHands):
    answer = INF
    n = len(clockHands)
    
    def rotate(y, x, t, table):
        for k in range(5):
            ny, nx = y+dy[k], x+dx[k]
            if (0 <= ny < n) and (0 <= nx < n):
                table[ny][nx] = (table[ny][nx]+t)%4
    
    for f in product(range(4), repeat=n):
        table = [i[:] for i in clockHands]
        for x in range(n):
            rotate(0, x, f[x], table)
        cnt = sum(f)
        
        for y in range(1, n):
            for x in range(n):
                if table[y-1][x]:
                    tmp = 4-table[y-1][x]
                    rotate(y, x, tmp, table)
                    cnt += tmp
        
        if sum(table[n-1]) == 0:
            answer = min(answer, cnt)
                
    return answer


clockHands = [[0,3,3,0],[3,2,2,3],[0,3,2,0],[0,3,3,3]]
print(solution(clockHands)) # 3

File: Java/programmers/etc/level3/고고학최고의발견.java
package CodeTest.Java.programmers.etc.level3;

public class 고고학최고의발견 {
    private static int N;
    private static int answer = Integer.MAX_VALUE;
    private static int[][] map;
    private static int[] rot;
    private static int[] dx = {0,0,0,1,-1};
    private static int[] dy = {0,1,-1,0,0};

    public int solution(int[][] clockHands) {
        N = clockHands.length;
        map = new int[N][N];
        rot = new int[N];
        
        for (int i=0; i<N; i++) {
            for (int j=0; j<N; j++) {
                map[i][j] = (4-clockHands[i][j])%4;
            }
        }

        dfs(0);
        return answer;
    }

    private static void dfs(int idx) {
        if (idx == N) {
            int[][] tmp = new int[N][N];
            int[] cur = new int[N];
            
            for (int i=0; i<N; i++) {
                cur[i] = rot[i];
                for (int j=0; j<N; j++) {
                    tmp[i][j] = map[i][j];
                }
            }

            int res = 0;
            for (int i=0; i<N; i++) {
                for (int j=0; j<N; j++) {
                    res += cur[j];
                    for (int d=0; d<5; d++) {
                        int nx = i+dx[d], ny = j+dy[d];
                        if (nx < 0 || nx >= N || ny < 0 || ny >= N) continue;
                        tmp[nx][ny] = (tmp[nx][ny]-cur[j]+4)%4;
                    }
                }
                for (int j=0; j<N; j++) {
                    cur[j] = tmp[i][j];
                }
            }

            boolean flag = true;
            for (int i=0; i<N; i++) {
                if (cur[i] != 0) {
                    flag = false;
                    break;
                }
            }
            
            if (flag) answer = Math.min(answer, res);
            return;
        }

        for (int i=0; i<4; i++) {
            rot[idx] = i;
            dfs(idx+1);
        }
    }
}
